xdsl/dialects/builtin.py

```diff
@@ -183,14 +166,83 @@ def print_parameter(self, printer: Printer) -> None:
     @staticmethod
     def generic_constraint_coercion(args: tuple[Any]) -> AttrConstraint:
         assert len(args) == 1
-        return ArrayOfConstraint(irdl_to_attr_constraint(args[0]))
+        return ArrayOfConstraint(RangeOf(irdl_to_attr_constraint(args[0])))
 
     def __len__(self):
         return len(self.data)
 
     def __iter__(self) -> Iterator[AttributeCovT]:
         return iter(self.data)
 
+    @staticmethod
+    def constr(
```

Comment:
I've found it's sometimes nicer to have invariant typevars in static methods, though to be honest don't understand the whole system fully enough to say if it's the "right" thing to do

Reply:
Both should work, right? Or do you see a case where this would fail?

Reply:
I feel, at least at some point, that this style would sometimes force you to write `ArrayAttr[PutTypeHere].constr(PutTypeHere)` whereas the invariant version wouldn't

Reply:
Ohhhhh, but wait, that's different then right?
We still want to use `CovT`, but just another version of it right? Like a `CovT2` so then it's unrelated to the actual generic type?

Reply:
Maybe that's also fine

Reply:
Wait actually, I think you are right and `InvT` works as well as `CovT`

Reply:
Yeah, my bad I'm dumb, I don't think `InvT` makes any sense in that context!

Reply:
It likely doesn't make a difference since the whole datatype is covariant anyway

Reply:
Should be fixed now!